src/io/image_reader_type.rs

```diff
@@ -354,10 +374,85 @@ impl ImageReader<BufReader<File>> {
             .filter(|ext| !ext.is_empty())
             .map(|ext| Format::Extension(ext.to_owned()));
 
-        Ok(ImageReader {
+        Ok(ImageFile {
             inner: BufReader::new(File::open(path)?),
             format,
             limits: Limits::default(),
         })
     }
 }
+
+impl ImageReader<'_> {
+    /// Decode the next image into a `DynamicImage`.
+    pub fn decode(&mut self) -> ImageResult<DynamicImage> {
+        // Try to resolve the viewbox via the decoder, fallback to `crop` if that is not possible.
+        let residual_vb = self.viewbox.and_then(|vb| self.inner.viewbox(vb).err());
+
+        let layout = self.inner.init()?;
+        // This is technically redundant but it's also cheap.
+        self.limits.check_dimensions(layout.width, layout.height)?;
+        // Check that we do not allocate a bigger buffer than we are allowed to
+        // FIXME: should this rather go in `DynamicImage::from_decoder` somehow?
+        self.limits.reserve(layout.total_bytes())?;
```

Comment:
This has a subtle behavior change. Previously we reserved the space for the output buffer and _then_ passed the reduced memory limit onto the decoder to limit its metadata/scratch buffer allocations. But at this point in the code we've already told the decoder that it is free to use the entire memory limit amount for those purposes. So we effectively only limiting memory to 2x the requested amount.